src/App.tsx

```diff
@@ -1,24 +1,280 @@
+      .catch((error: Error) => {
+        setErrorMessage(ErrorMessage.UnableLoad);
+        // eslint-disable-next-line no-console
+        console.log(error.message);
+      });
+  }, []);
+
+  const handleTodoAdd = (newTodoTitle: string) => {
+    if (!newTodoTitle.trim()) {
+      setErrorMessage(ErrorMessage.EmptyTitle);
+
+      return;
+    }
+
+    setTempTodo({
+      id: 0,
+      title: newTodoTitle,
+      userId: 0,
+      completed: false,
+    });
+
+    // eslint-disable-next-line consistent-return
+    return ApiServices
+      .addTodo(newTodoTitle.trim())
+      .then((newTodo: Todo) => {
+        setTodos(prevTodos => [...prevTodos, newTodo]);
+      }).catch(error => {
+        setErrorMessage(ErrorMessage.UnableAdd);
+        throw error;
+      }).finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const handleTodoDelete = (todoId: number) => {
+    setProcessingTodoIds(prevState => [...prevState, todoId]);
+
+    return ApiServices
+      .deleteTodo(todoId)
+      .then(() => {
+        setErrorMessage('');
+        setTodos((prevTodos: Todo[]) => (
+          prevTodos.filter(todo => todo.id !== todoId)
+        ));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.UnableDelete);
+      })
+      .finally(() => {
+        setProcessingTodoIds(prevState => prevState.filter(id => id !== todoId));
+      });
+  };
+
+  const handleTodoRename = (todo: Todo, newTodoTitle: string) => {
+    setProcessingTodoIds(prevState => [...prevState, todo.id]);
+
+    ApiServices
+      .updateTodo({
+        id: todo.id,
+        title: newTodoTitle,
+        userId: todo.userId,
+        completed: todo.completed,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map((currentTodo: Todo) => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(error => {
+        setErrorMessage(ErrorMessage.UnableUpdate);
+        throw error;
+      })
+      .finally(() => {
+        setProcessingTodoIds(prevState => prevState.filter(id => id !== todo.id));
+      });
+  };
+
+  const handleTodoToggle = (todo: Todo) => {
+    setProcessingTodoIds(prevState => [...prevState, todo.id]);
+
+    ApiServices
+      .updateTodo({
+        ...todo,
+        completed: !todo.completed,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map((currentTodo: Todo) => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(error => {
+        setErrorMessage(ErrorMessage.UnableUpdate);
+        throw error;
+      })
+      .finally(() => {
+        setProcessingTodoIds(prevState => prevState.filter(id => id !== todo.id));
+      });
+  };
+
+  const handleClearCompletedTodos = () => {
+    todos
+      .filter(todo => todo.completed)
+      .forEach(todo => handleTodoDelete(todo.id));
```

Comment:
No need to do extra filtering, this condition can be transferred to the middle of the `forEach` method